toolchain: Add the  __deprecated_version macro

rruuaanng opened the pull request:
For more to see #87662

Closes: #87662

stephanosio commented on include/zephyr/toolchain.h:
```diff
@@ -318,6 +318,18 @@
+ *
+ * @param removal_version Version of planned removal
+ */
+#if defined(__GNUC__) || defined(__llvm__)
+#define TOOLCHAIN_DEPRECATED(removal_version) __deprecated_version(removal_version)
+#else
+#define TOOLCHAIN_DEPRECATED(removal_version)
+#endif
```
This is pretending to be one of the toolchain warning control macros when it is not -- `TOOLCHAIN_DEPRECATED` is unnecessary and should be removed.

Only `__deprecated_version` should be used in code and that should be made available in every toolchain supported.

stephanosio commented on include/zephyr/toolchain/gcc.h:
```diff
@@ -267,6 +267,10 @@ do {                                                                    \
  */
 #endif
 
+#ifndef __deprecated_version
+#define __deprecated_version(version)	__attribute__((deprecated(#version)))
+#endif
```
We need to ensure that this macro is available in every toolchain supported (e.g. `iccarm.h` included by `iar.h`); otherwise, code using this macro will fail to compile on the toolchains missing this macro.

stephanosio reviewed (changes requested):
Also, the warning message output by this macro is rather cryptic:

```
test.c:19:5: warning: 'foo' is deprecated: 4.3 [-Wdeprecated-declarations]
```

The message should be more clear to say something like "planned removal in v4.3"

stephanosio commented on include/zephyr/toolchain/iar/iccarm.h:
```diff
@@ -185,6 +185,11 @@ do {                                                                    \
 #define __deprecated	__attribute__((deprecated))
 #endif
 
+#ifndef __deprecated_version
+#define __deprecated_version(version) \
+	__attribute__((deprecated("planned removal in v" #version)))
+#endif
```
@RobinKastberg @bjorniuppsala @LoveKarlsson Could someone from IAR confirm that the `deprecated` attribute with `(msg)` is supported in IAR?

stephanosio reviewed (commented):
Need to update `checkpatch` attribute list:

https://github.com/zephyrproject-rtos/zephyr/blob/4cae4661a262215ee288041371b4daa84dfabc9e/scripts/checkpatch.pl#L379

And would be nice to update `tags.sh` as well so that these show up in the same way as `__deprecated`:

https://github.com/zephyrproject-rtos/zephyr/blob/4cae4661a262215ee288041371b4daa84dfabc9e/scripts/tags.sh#L61

stephanosio commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -116,6 +116,10 @@ New APIs and options
   * :kconfig:option:`CONFIG_SRAM_VECTOR_TABLE` moved from ``zephyr/Kconfig.zephyr`` to
     ``zephyr/arch/Kconfig`` and added dependencies to it.
 
+* Toolchain
+
+  * :c:macro:`__deprecated_version`
+
```
This will need to go into 4.3 since we are already past the feature freeze for 4.2.

rruuaanng committed:
toolchain: gcc: Add the __deprecated_version macro

Add the `__deprecated_version` macro to print version
information when deprecation warnings are triggered.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
toolchain: iar: Add the __deprecated_version macro

Add the `__deprecated_version` macro to print version
information when deprecation warnings are triggered.

Signed-off-by: James Roy <rruuaanng@outlook.com>

dkalowsk commented:
@rruuaanng you have a conflict on the release-notes.  Can you please fix?

rruuaanng committed:
scripts: Add __deprecated_version to the checkpatch checklist

Add macro __deprecated_version to the checklists of
`scripts/checkpatch.pl` and `scripts/tags.sh`.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
doc: release: Add a release entry

Add `__deprecated_version` macro to the Toolchain entry list.

Signed-off-by: James Roy <rruuaanng@outlook.com>
